README.md
# README

![alt tag](https://raw.githubusercontent.com/rbern/debroot/master/debroot.png)

An app that can:

* create (debootstrap) a debian or ubuntu system;
* run a shell in the chroot filesystem;
* create a live iso from the created chroot filesystem;
* unsquash, update and rebuild an existing ubuntu or debian and their derivatives live ISO image.

ATM the app does the following:

* selects and unsqhashes an Debian or Ubuntu Live iso file (content of chroot may be upgraded, see bellow) and can rebuild it;
* chroot APT sources.list can be edited, "apt-get update/upgrade/dist-upgrade" are a button click away;
* packages can be installed in the chroot by dropping their names in a textbox and clicking "Install";
* can create a bare minbase debootstrap of Ubuntu or Debian, and from that, create a very small Live ISO, or else one can add new repositories, install packages and then build the Live iso image;
* generates a script with all the commands that were executed to replicate the process on another host without debroot.

# Bugs

* To enable "Unsquash" button you may have to type and remove text in the "ROOTFS Directory" text input;
* Some buttons are disabled because they are not implemeted yet. It is the case of "build USB image" and "chroot RDP" buttons.

Also see TODO file.

# ROADMAP

* run scripts/hooks in chroot;
* support preseed files upon package install;
* build non-live USB image;
* add a custom script based installer to "replace" both ubiquity or debian-installer;

# Install

There are pre-built debian packages available at <https://sourceforge.net/projects/debroot/>.

One can also build a debian package from sources with:

	dpkg-buildpackage -us -uc

## Install from debian package

Install it with gdebi:

	sudo gdebi -o "APT::Install-Recommends=0" debroot_0.XX_all.deb -nq

### Execute

From an installed debian package:

	sudo debroot.pl

## Install from sources

Unzip or git clone somewhere, then...

### Runtime dependencies

This is a perl QT app. To run it in debian/Ubuntu install:

	sudo apt-get install libqtgui4-perl

### Execute

From sources:

	sudo ./debroot.pl

### Development time dependencies

To develop install:

	sudo apt-get install qt4-dev-tools

To update the GUI perl module after changes made in qt4-designer type:

	make

# How to

[Forum thread in debian.forums.net](http://forums.debian.net/viewtopic.php?f=16&t=128995);

[Forum thread in ubuntuforums.com](https://ubuntuforums.org/showthread.php?t=2330142);

## Quick build a clean live iso from stratch

* Select the distribution (Ubuntu or Debian) and select the release;
* If you have selected Ubuntu, on the Debootstrap tab, on the "Additional options" type "--components=main,universe" to allow to download some packages from there;
* Type the directory where the new chroot will be built in "ROOTFS directory";
* Click "Debootstrap" button and wait until it finishes;
* Click the "Build" tab. You will have a "Prepare ISO" button available while "Build ISO" is not. Click "Prepare ISO" and wait. Reply "Yes" to all package installs needed to make a Live system;
* After "Prepare ISO" is finished, the button "Build ISO" should now be available. Click it to build the Live iso of a chroot.

## Upgrade and rebuild of an existing Live ISO

* Type the directory where the ISO chroot content will be extrated in "ROOTFS directory";
* Click the "Unsquash" tab, click "Select ISO" button to select an existing ISO;
* After the file is selected the button "Unsquash" should now be available. If not add and remove text in the "ROOTFS directory" to update the UI interface (its a bug);
* Click the "Unsquash" button and wait;
* Click the "sources.list/upgrade" tab. Click the button "Read" to load the existing "sources.list" present in the chroot. Click the "Update" button to run an "apt-get update" in the chroot;
* Click the "Upgrade" button to upgrade the content of the chroot, as a "apt-get upgrade" would do. Running "dist-upgrade" is not recommended, but available;
* Click the "Install" tab to install additional packages to the live image;
* After installing or upgrading the system, click the tab "Chroot" and then click the button "Chroot Shell";
* In the shell type "apt-get autoremove --purge" and "apt-get clean" to reduce the size of the resulting live image. Type "exit" to return to the app;
* Now click the "Build" tab and the button "Build ISO" should be available. Click it.

## Rebuild non-UEFI Debian live images with UEFI support

This is a workaround for Debian 7 Wheezy and 8 Jessie and testing/Stretch, while [bug #820036](http://bugs.debian.org/820036) is not fixed for Debian 9 Stretch. It is only supported for 64 bit systems.

More info about this UEFI workaround see [James Bottomley blog](http://blog.hansenpartnership.com/linux-foundation-secure-boot-system-released/) about it.

* Type the directory where the ISO chroot content will be extrated in "ROOTFS directory";
* Click the "Unsquash" tab, click "Select ISO" button to select an existing ISO;
* After the file is selected the button "Unsquash" should now be available. If not add and remove text in the "ROOTFS directory" to update the UI interface (its a bug);
* Click the "Unsquash" button and wait;
* Now click the "Build" tab and the button "Build ISO" should be available. Click it without unchecking the checkbox "UEFI enabled".

# Tested distros (Unsquash/upgrade/rebuild)

This app works with either ubuntu and debian and their derivatives. It was tested with the following distros:

* Ubuntu 16.04;
* Debian 7 and 8 (although Debian produces Live daily builds, one can install additional packages with debroot);
* Linux Mint (and LMDE2);
* Kali Linux 2016.1;
* TAILS 2.4;
* Sparkylinux 4.3;
* Debtailor 2016-1;
* In general this app should work with any live distro that uses either casper or debian-live.

# Copyright

This project is licensed under the terms of the GNU GPL v2 License.

(c) 2016 Rui Miguel P. Bernardo

Email: rui.bernardo.pt(at)gmail.com


debian/changelog
debroot (0.11) unstable; urgency=medium

  * Run apt-get update before reinstalling kernel.
  * Check if lzma compression was used in image and rebuild image 
    accordingly (Fixes #1: Rebuilt LMDE 2 points to initrd.lz).
  * Update README file.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Wed, 21 Sep 2016 03:13:29 +0100

debroot (0.10) unstable; urgency=medium

  * Fix typo.
  * Fix i386 ubuntu live iso creation.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Tue, 06 Sep 2016 14:19:43 +0100

debroot (0.09) unstable; urgency=medium

  * Only create and remove devices in /dev (kali) if we created them.
  * When installing dbus, stop it.
  * Give more info to user about some steps in the build process.
  * Use same variable name in multiple places; less confusing.
  * Create save and read file functions.
  * Update README with new version and add links to forum threads.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Sat, 30 Jul 2016 18:27:34 +0100

debroot (0.08) unstable; urgency=medium

  * Fix dpkg query to generate manifest of the live iso, in both ubuntu 
    and debian.
  * Add a grub entry for debian live-installer with preseed from UEFI boot.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Mon, 18 Jul 2016 07:23:46 +0100

debroot (0.07) unstable; urgency=medium

  * Fix and update README.md.
  * Remove backports from Ubuntu recommended sources.list file.
  * When running apt-get update try to do it faster.
  * Fix syslinux-themes-ubuntu package selection.
  * Use correct iso check option on uefi grub configuration.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Sun, 10 Jul 2016 06:35:53 +0100

debroot (0.06) unstable; urgency=medium

  * Fix path to grub file when not running from sources.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Thu, 07 Jul 2016 18:18:22 +0100

debroot (0.05) unstable; urgency=medium

  * Support 64 bit UEFI live boot for created and rebuilt images.
  * Enable the possibily to not build isohybrid images (pure iso files).

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Thu, 07 Jul 2016 17:55:07 +0100

debroot (0.04) unstable; urgency=medium

  * Update README.md.
  * Dont remove UI perl module on clean. It's in git.
  * Fix paths of iso file if relative paths are used in ROOTFS directory 
    textbox.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Tue, 05 Jul 2016 01:34:56 +0100

debroot (0.03) unstable; urgency=medium

  * Add rebuilt file from last release.
  * Fix: always build isohybrid images.
  * When preparing the iso dont rely on what is selected but on what present is in chroot.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Mon, 04 Jul 2016 22:04:36 +0100

debroot (0.02) unstable; urgency=medium

  * Add install rules to Makefile, so debian packaging works.
  * Fix: the "prepare iso" button should be available if $dir-binary does not exist.
  * Cleanup $dir/dev/{random,urandom,null} that were created with mknod.
  * Fix help in "Install" tab: sources.list text in messagebox has copy/paste.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Mon, 04 Jul 2016 17:28:20 +0100

debroot (0.01) unstable; urgency=low

  * Initial Release.

 -- Rui Miguel P. Bernardo <rui.bernardo.pt@gmail.com>  Sat, 02 Jul 2016 00:47:49 +0100


fake-grub-probe.sh
#!/bin/bash
#
# A faker script for grub-probe, if your's stops working properly
#

# Depends on /root/grub.hack file

# . /root/grub.hach

arg_is_device=0
target=""

# Call Order:
# / --target=device
# --target=fs_uuid /dev/sda1 --device
# /boot --target=device
# --target=fs_uuid /boot --device
# "/" "--target=fs"
# "--target=abstraction" "ext2" "--device"
# "--target=fs" "ext2" "--device"
# "--target=drive" "ext2" "--device"
# "--target=fs_uuid" "ext2" "--device"
# "--target=abstraction" "ext2" "--device"
# "--target=fs" "ext2" "--device"
# "--target=drive" "ext2" "--device"
# "--target=fs_uuid" "ext2" "--device"
# "/" "--target=device"
# "--target=fs_uuid" "ext2" "--device"
# "/boot" "--target=device"
# "--target=fs_uuid" "ext2" "--device"
# "/" "--target=fs"
# "--target=abstraction" "ext2" "--device"
# "--target=fs" "ext2" "--device"
# "--target=drive" "ext2" "--device"
# "--target=fs_uuid" "ext2" "--device"
# "--target=abstraction" "ext2" "--device"
# "--target=fs" "ext2" "--device"
# "--target=drive" "ext2" "--device"
# "--target=fs_uuid" "ext2" "--device"

for opt in $@
do
    case "${opt}" in
    --device)
        # Means the thing in $arg is a device
        arg_is_device=1
        ;;
    --target=*)
        # = device, = fs, = fs_uuid
        target=${opt#*=}
        ;;
    *)
        arg=$opt
        ;;
    esac
done

set >> /tmp/probe.env

case "$target" in
abstraction)
    echo
    ;;
device)
    echo "/dev/sda1"
    ;;
drive)
    echo "(hd0)"
    ;;
fs)
    # Returns the filesystem type
    echo "ext2"
    ;;
fs_uuid)
    tune2fs -l $arg | awk '/Filesystem UUID/ { print $3 }'
    ;;
esac



Ui_DebootForm.pm
###############################################################################
#
###############################################################################

#################################################################################
## Form generated from reading UI file 'debrootform.ui'
##
## Created: ter Jun 14 23:05:17 2016
##      by: Qt User Interface Compiler version 4.8.6
##
## WARNING! All changes made in this file will be lost when recompiling UI file!
################################################################################

package Ui_Form;

use strict;
use warnings;
use QtCore4;
use QtGui4;

sub groupBox {
    return shift->{groupBox};
}

sub radioButtonDebian {
    return shift->{radioButtonDebian};
}

sub radioButtonUbuntu {
    return shift->{radioButtonUbuntu};
}

sub groupBox_2 {
    return shift->{groupBox_2};
}

sub comboBox {
    return shift->{comboBox};
}

sub tabWidget {
    return shift->{tabWidget};
}

sub tab {
    return shift->{tab};
}

sub pushButtonDebootstrap {
    return shift->{pushButtonDebootstrap};
}

sub label_2 {
    return shift->{label_2};
}

sub lineEditMirror {
    return shift->{lineEditMirror};
}

sub tab_2 {
    return shift->{tab_2};
}

sub tab_3 {
    return shift->{tab_3};
}

sub tab_4 {
    return shift->{tab_4};
}

sub label {
    return shift->{label};
}

sub lineEditDirectory {
    return shift->{lineEditDirectory};
}


sub setupUi {
    my ( $class, $form ) = @_;
    my $self = bless {}, $class;
    if ( !defined $form->objectName() ) {
        $form->setObjectName( "form" );
    }
    $form->resize( 400, 300 );
    my $groupBox = Qt::GroupBox( $form );
    $self->{groupBox} = $groupBox;
    $groupBox->setObjectName( "groupBox" );
    $groupBox->setGeometry( Qt::Rect(40, 30, 120, 80) );
    my $radioButtonDebian = Qt::RadioButton( $groupBox );
    $self->{radioButtonDebian} = $radioButtonDebian;
    $radioButtonDebian->setObjectName( "radioButtonDebian" );
    $radioButtonDebian->setGeometry( Qt::Rect(10, 20, 105, 20) );
    my $radioButtonUbuntu = Qt::RadioButton( $groupBox );
    $self->{radioButtonUbuntu} = $radioButtonUbuntu;
    $radioButtonUbuntu->setObjectName( "radioButtonUbuntu" );
    $radioButtonUbuntu->setGeometry( Qt::Rect(10, 50, 105, 20) );
    my $groupBox_2 = Qt::GroupBox( $form );
    $self->{groupBox_2} = $groupBox_2;
    $groupBox_2->setObjectName( "groupBox_2" );
    $groupBox_2->setGeometry( Qt::Rect(220, 30, 120, 101) );
    my $comboBox = Qt::ComboBox( $groupBox_2 );
    $self->{comboBox} = $comboBox;
    $comboBox->setObjectName( "comboBox" );
    $comboBox->setGeometry( Qt::Rect(10, 20, 101, 25) );
    my $tabWidget = Qt::TabWidget( $form );
    $self->{tabWidget} = $tabWidget;
    $tabWidget->setObjectName( "tabWidget" );
    $tabWidget->setGeometry( Qt::Rect(10, 140, 371, 151) );
    my $tab = Qt::Widget(  );
    $self->{tab} = $tab;
    $tab->setObjectName( "tab" );
    my $pushButtonDebootstrap = Qt::PushButton( $tab );
    $self->{pushButtonDebootstrap} = $pushButtonDebootstrap;
    $pushButtonDebootstrap->setObjectName( "pushButtonDebootstrap" );
    $pushButtonDebootstrap->setGeometry( Qt::Rect(260, 80, 92, 27) );
    my $label_2 = Qt::Label( $tab );
    $self->{label_2} = $label_2;
    $label_2->setObjectName( "label_2" );
    $label_2->setGeometry( Qt::Rect(10, 10, 41, 16) );
    my $lineEditMirror = Qt::LineEdit( $tab );
    $self->{lineEditMirror} = $lineEditMirror;
    $lineEditMirror->setObjectName( "lineEditMirror" );
    $lineEditMirror->setGeometry( Qt::Rect(10, 40, 341, 25) );
    $tabWidget->addTab( $tab, Qt::Application::translate( 'Form', "Debootstrap", undef, Qt::Application::UnicodeUTF8() ) );
    my $tab_2 = Qt::Widget(  );
    $self->{tab_2} = $tab_2;
    $tab_2->setObjectName( "tab_2" );
    $tabWidget->addTab( $tab_2, Qt::Application::translate( 'Form', "sources.list/upgrade", undef, Qt::Application::UnicodeUTF8() ) );
    my $tab_3 = Qt::Widget(  );
    $self->{tab_3} = $tab_3;
    $tab_3->setObjectName( "tab_3" );
    $tabWidget->addTab( $tab_3, Qt::Application::translate( 'Form', "Install", undef, Qt::Application::UnicodeUTF8() ) );
    my $tab_4 = Qt::Widget(  );
    $self->{tab_4} = $tab_4;
    $tab_4->setObjectName( "tab_4" );
    $tabWidget->addTab( $tab_4, Qt::Application::translate( 'Form', "Chroot", undef, Qt::Application::UnicodeUTF8() ) );
    my $label = Qt::Label( $form );
    $self->{label} = $label;
    $label->setObjectName( "label" );
    $label->setGeometry( Qt::Rect(20, 110, 61, 20) );
    my $lineEditDirectory = Qt::LineEdit( $form );
    $self->{lineEditDirectory} = $lineEditDirectory;
    $lineEditDirectory->setObjectName( "lineEditDirectory" );
    $lineEditDirectory->setGeometry( Qt::Rect(100, 110, 281, 25) );

    $self->retranslateUi( $form );

    $tabWidget->setCurrentIndex( 0 );


    Qt::MetaObject::connectSlotsByName( $form );
    return $self;
} # setupUi

sub setup_ui {
    my ( $form ) = @_;
    return setupUi( $form );
}

sub retranslateUi {
    my ( $self, $form ) = @_;
    $form->setWindowTitle( Qt::Application::translate( 'Form', "debroot", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{groupBox}->setTitle( Qt::Application::translate( 'Form', "Distribution", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{radioButtonDebian}->setText( Qt::Application::translate( 'Form', "Debian", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{radioButtonUbuntu}->setText( Qt::Application::translate( 'Form', "Ubuntu", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{groupBox_2}->setTitle( Qt::Application::translate( 'Form', "Release", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{pushButtonDebootstrap}->setText( Qt::Application::translate( 'Form', "debootstrap", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{label_2}->setText( Qt::Application::translate( 'Form', "Mirror:", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{tabWidget}->setTabText( $self->{tabWidget}->indexOf( $self->{tab}), Qt::Application::translate( 'Form', "Debootstrap", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{tabWidget}->setTabText( $self->{tabWidget}->indexOf( $self->{tab_2}), Qt::Application::translate( 'Form', "sources.list/upgrade", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{tabWidget}->setTabText( $self->{tabWidget}->indexOf( $self->{tab_3}), Qt::Application::translate( 'Form', "Install", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{tabWidget}->setTabText( $self->{tabWidget}->indexOf( $self->{tab_4}), Qt::Application::translate( 'Form', "Chroot", undef, Qt::Application::UnicodeUTF8() ) );
    $self->{label}->setText( Qt::Application::translate( 'Form', "Directory:", undef, Qt::Application::UnicodeUTF8() ) );
} # retranslateUi

sub retranslate_ui {
    my ( $form ) = @_;
    retranslateUi( $form );
}

1;